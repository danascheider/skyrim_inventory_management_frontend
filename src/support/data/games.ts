import { ResponseGame as Game } from '../../types/apiData'

export const emptyGames: Game[] = []

export const allGames: Game[] = [
  {
    id: 32,
    user_id: 412,
    name: 'My Game 1',
    description: 'This is a game with a description',
    created_at: new Date(),
    updated_at: new Date(),
  },
  {
    id: 51,
    user_id: 412,
    name: 'My Game 2',
    description: null,
    created_at: new Date(),
    updated_at: new Date(),
  },
  {
    id: 77,
    user_id: 412,
    name: 'Game with a really really really really really long name',
    description:
      'Cum audissem Antiochum, Brute, ut solebam, cum M. Pisone in eo gymnasio, quod Ptolomaeum vocatur, unaque nobiscum Q. frater et T. Pomponius postmeridianam conficeremus in Academia',
    created_at: new Date(),
    updated_at: new Date(),
  },
]
