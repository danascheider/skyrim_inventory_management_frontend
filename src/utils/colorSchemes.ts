export interface ColorScheme {
  schemeColorDarkest: string
  schemeColorDark: string
  schemeColorLight: string
  schemeColorLightest: string
  hoverColorDark: string
  hoverColorLight: string
  hoverColorLightest: string
  textColorPrimary: string
  textColorSecondary: string
  textColorTertiary: string
  borderColor: string
}

export const YELLOW: ColorScheme = {
  schemeColorDarkest: '#ffbf00',
  schemeColorDark: '#ffcb32',
  schemeColorLight: '#ffd866',
  schemeColorLightest: '#fff2cc',
  hoverColorDark: '#e5ab00',
  hoverColorLight: '#ffc519',
  hoverColorLightest: '#ffebb2',
  textColorPrimary: '#4c3900',
  textColorSecondary: '#4c3900',
  textColorTertiary: '#7f5700',
  borderColor: '#cc9800',
}

export const PINK: ColorScheme = {
  schemeColorDarkest: '#e83f6f',
  schemeColorDark: '#ec658b',
  schemeColorLight: '#f18ba8',
  schemeColorLightest: '#fad8e2',
  hoverColorDark: '#d03863',
  hoverColorLight: '#ea527d',
  hoverColorLightest: '#f5b2c5',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#451221',
  borderColor: '#b93258',
}

export const BLUE: ColorScheme = {
  schemeColorDarkest: '#2274a5',
  schemeColorDark: '#4e8fb7',
  schemeColorLight: '#7aabc9',
  schemeColorLightest: '#d2e3ed',
  hoverColorDark: '#1e6894',
  hoverColorLight: '#3881ae',
  hoverColorLightest: '#bcd5e4',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#0d2e42',
  borderColor: '#1b5c84',
}

export const GREEN: ColorScheme = {
  schemeColorDarkest: '#00a323',
  schemeColorDark: '#32b54e',
  schemeColorLight: '#66c77b',
  schemeColorLightest: '#ccecd3',
  hoverColorDark: '#00921f',
  hoverColorLight: '#19ac38',
  hoverColorLightest: '#b2e3bd',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#00410e',
  borderColor: '#00821c',
}

export const AQUA: ColorScheme = {
  schemeColorDarkest: '#20e2e9',
  schemeColorDark: '#62eaef',
  schemeColorLight: '#8ff0f4',
  schemeColorLightest: '#d2f9fa',
  hoverColorDark: '#11cbd1',
  hoverColorLight: '#4ce7ed',
  hoverColorLightest: '#bcf6f8',
  textColorPrimary: '#094345',
  textColorSecondary: '#094345',
  textColorTertiary: '#094345',
  borderColor: '#19b4ba',
}

const colorSchemes: ColorScheme[] = [YELLOW, PINK, BLUE, GREEN, AQUA]

export default colorSchemes
