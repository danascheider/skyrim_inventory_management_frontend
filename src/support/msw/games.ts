import { rest } from 'msw'
import { ResponseGame } from '../../types/apiData'
import { emptyGames, allGames } from '../data/games'

const BASE_URI = 'http://localhost:3000'

/**
 *
 * POST /games
 *
 */

export const postGamesSuccess = rest.post(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    const body = {
      id: 102,
      user_id: 412,
      name: req.params.name || 'My Game 3',
      description:
        req.params.description || 'This description is just for illustration.',
      created_at: new Date(),
      updated_at: new Date(),
    }

    return res(ctx.status(201), ctx.json(body))
  }
)

export const postGamesUnprocessable = rest.post(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    const body = {
      errors: [
        "Name can only contain alphanumeric characters, spaces, commas (,), hyphens (-), and apostrophes (')",
      ],
    }

    return res(ctx.status(422), ctx.json(body))
  }
)

export const postGamesServerError = rest.post(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    const body = { errors: ['oh noes'] }

    return res(ctx.status(500), ctx.json(body))
  }
)

/**
 *
 * GET /games
 *
 */

export const getGamesEmptySuccess = rest.get(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(emptyGames))
  }
)

export const getGamesAllSuccess = rest.get(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(allGames))
  }
)

export const getGamesServerError = rest.get(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    return res(ctx.status(500), ctx.json({ errors: ['Something went wrong'] }))
  }
)

/**
 *
 * PATCH /games/:id
 *
 */

const newOrExistingGame = (id: number): ResponseGame => {
  const existingGame = allGames.find((game) => game.id === id)

  if (existingGame) return existingGame

  return {
    id,
    user_id: 412,
    name: 'Skyrim Game',
    description: null,
    created_at: new Date(),
    updated_at: new Date()
  }
}

export const patchGameSuccess = rest.patch(
  `${BASE_URI}/games/:id`,
  async (req, res, ctx) => {
    const id = Number(req.params.id)
    const body = await req.json()
    const game = newOrExistingGame(id)

    return res(
      ctx.status(200),
      ctx.json({ ...game, ...body })
    )
  }
)

export const patchGameUnprocessableEntity = rest.patch(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
  return res(
    ctx.status(422),
    ctx.json({ errors: ['Name must be unique'] })
  )
})

export const patchGameNotFound = rest.patch(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(
      ctx.status(404)
    )
  }
)

export const patchGameServerError = rest.patch(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({ errors: ['oh noes'] })
    )
  }
)

/**
 *
 * DELETE /games/:id
 *
 */

export const deleteGameSuccess = rest.delete(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(ctx.status(204))
  }
)

export const deleteGameNotFound = rest.delete(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(ctx.status(404))
  }
)

export const deleteGameServerError = rest.delete(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(ctx.status(500), ctx.json({ errors: ['oh noes'] }))
  }
)
