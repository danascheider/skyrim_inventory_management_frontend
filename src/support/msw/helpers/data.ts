import {
  type RequestShoppingListItem,
  type RequestShoppingList,
  type ResponseShoppingListItem,
  type ResponseShoppingList,
} from '../../../types/apiData'
import { allGames } from '../../data/games'
import {
  allShoppingLists,
  shoppingListsForGame,
} from '../../data/shoppingLists'

const gameIds = allGames.map(({ id }) => id)

/**
 *
 * Shopping List Creation
 *
 */

export const newShoppingList = (
  attributes: RequestShoppingList,
  gameId: number
): ResponseShoppingList[] => {
  if (gameIds.indexOf(gameId) < 0)
    throw new Error(
      'Cannot generate shopping list for game that does not exist in test data'
    )

  const existingLists = shoppingListsForGame(gameId)

  if (!existingLists.length)
    throw new Error(
      'Cannot generate single list for game without existing aggregate'
    )

  return [
    {
      id: 93,
      game_id: gameId,
      aggregate: false,
      aggregate_list_id: existingLists[0].id,
      title: attributes.title || 'New Shopping List',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
  ]
}

export const newShoppingListWithAggregate = (
  attributes: RequestShoppingList,
  gameId: number
): ResponseShoppingList[] => {
  if (gameIds.indexOf(gameId) < 0)
    throw new Error(
      'Cannot generate shopping list for game that does not exist in test data'
    )

  const existingLists = shoppingListsForGame(gameId)

  if (existingLists.length)
    throw new Error(
      'Cannot generate new aggregate list for game that already has one'
    )

  return [
    {
      id: 93,
      game_id: gameId,
      aggregate: true,
      aggregate_list_id: null,
      title: 'All Items',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
    {
      id: 94,
      game_id: gameId,
      aggregate: false,
      aggregate_list_id: 93,
      title: attributes.title || 'My Shopping List 1',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
  ]
}

/**
 *
 * Shopping list item creation
 *
 */

export const newShoppingListItem = (
  attributes: RequestShoppingListItem,
  listId: number
) => {
  const list = allShoppingLists.find(({ id }) => id === listId)

  if (!list)
    throw new Error(`No shopping list with ID ${listId} in the test data`)

  const shoppingList = { ...list }
  const allLists = shoppingListsForGame(shoppingList.game_id)
  const aggregateList = { ...allLists[0] }

  const newItem: ResponseShoppingListItem = {
    id: 42,
    list_id: listId,
    description: 'Dummy description for TypeScript',
    quantity: 1,
    unit_weight: null,
    notes: null,
    created_at: new Date(),
    updated_at: new Date(),
    ...attributes,
  }

  const newAggregateListItem: ResponseShoppingListItem = {
    ...newItem,
    id: 43,
    list_id: aggregateList.id,
  }

  aggregateList.list_items = [newAggregateListItem, ...aggregateList.list_items]
  shoppingList.list_items = [newItem, ...shoppingList.list_items]

  return [aggregateList, shoppingList]
}
