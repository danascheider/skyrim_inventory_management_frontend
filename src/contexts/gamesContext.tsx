import { createContext, useCallback, useEffect, useState } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type Game } from '../types/games'
import { type ProviderProps } from '../types/contexts'
import { useGoogleLogin, usePageContext } from '../hooks/contexts'
import { ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import { getGames, deleteGame } from '../utils/simApi'

// TODO: Add default values for context provider as follows:
//       const createGame = (game: Game) => {
//                                            postGames(game: Game, token: '')
//                                              .then((status: number, json: string) => /* noop */)
//                                          }
//       This will be useful in Storybook.

const NOT_FOUND_MESSAGE =
  "Oops! We couldn't find the game you're looking for. Please refresh and try again."
const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface GamesContextType {
  games: Game[]
  gamesLoadingState: LoadingState
  // createGame: (game: Game) => void
  // updateGame: (gameId: number, attrs: Game) => void
  destroyGame: (gameId: number) => void
}

export const GamesContext = createContext<GamesContextType>({
  games: [],
  gamesLoadingState: LOADING,
  destroyGame: () => {},
})

export const GamesProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const [gamesLoadingState, setGamesLoadingState] = useState(LOADING)
  const [games, setGames] = useState<Game[]>([])
  const { setFlashProps } = usePageContext()

  /*
   *
   * Retrieve all the current user's games from the API
   *
   */

  const fetchGames = useCallback(() => {
    if (user && token) {
      getGames(token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setGames(json)
            setGamesLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (import.meta.env.DEV)
            console.error(`Error ${e.name}: ${e.message}`)

          if (e.code === 401) {
            signOutWithGoogle()
          } else {
            setGames([])
            setGamesLoadingState(ERROR)
            setFlashProps({
              type: 'error',
              message: e.message,
              hidden: false,
            })
          }
        })
    }
  }, [user, token])

  /*
   *
   * Destroy the requested game and update the `games` array
   *
   */

  const destroyGame = useCallback(
    (gameId: number) => {
      if (user && token) {
        deleteGame(gameId, token)
          .then(({ status }) => {
            if (status === 204) {
              const newGames = games.filter(({ id }) => id !== gameId)
              setGames(newGames)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your game has been deleted.',
              })
            }
          })
          .catch((e: ApiError) => {
            if (import.meta.env.DEV)
              console.log(`Error ${e.name}: ${e.message}`)

            const message =
              e.code === 404 ? NOT_FOUND_MESSAGE : UNEXPECTED_ERROR_MESSAGE

            setFlashProps({
              hidden: false,
              type: 'error',
              message,
            })
          })
      }
    },
    [user, token, games]
  )

  const value = {
    games,
    gamesLoadingState,
    destroyGame,
  }

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  useEffect(() => {
    if (authLoading) return

    fetchGames()
  }, [authLoading, fetchGames])

  return <GamesContext.Provider value={value}>{children}</GamesContext.Provider>
}
