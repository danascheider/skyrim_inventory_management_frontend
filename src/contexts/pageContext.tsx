import { createContext, useState, useEffect, useRef } from 'react'
import { type ApiCalls, type Resource, type HttpVerb } from '../types/apiCalls'
import { ProviderProps } from '../types/contexts'
import { type FlashProps, type ModalProps } from '../types/pageContext'

interface PageContextType {
  flashProps: FlashProps
  setFlashProps: (props: FlashProps) => void
  modalProps: ModalProps
  setModalProps: (props: ModalProps) => void
  apiCallsInProgress: ApiCalls
  addApiCall: (key: Resource, value: HttpVerb) => void
  removeApiCall: (key: Resource, value: HttpVerb) => void
}

const defaultFlashProps: FlashProps = {
  type: 'info',
  message: '',
  hidden: true,
}

const defaultModalProps: ModalProps = {
  hidden: true,
  children: <></>,
}

const defaultApiCallStatus = {
  games: [],
  shoppingLists: [],
  shoppingListItems: [],
}

const PageContext = createContext<PageContextType>({
  flashProps: defaultFlashProps,
  modalProps: defaultModalProps,
  apiCallsInProgress: defaultApiCallStatus,
  setFlashProps: () => {},
  setModalProps: () => {},
  addApiCall: () => {},
  removeApiCall: () => {},
})

const PageProvider = ({ children }: ProviderProps) => {
  const [flashProps, setFlashProps] = useState<FlashProps>(defaultFlashProps)
  const [modalProps, setModalProps] = useState<ModalProps>(defaultModalProps)
  const [apiCallsInProgress, SetApiCallsInProgress] =
    useState<ApiCalls>(defaultApiCallStatus)

  const flashVisibleSince = useRef(0)

  const addApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: [...apiCallsInProgress[key], value],
    })
  }

  const removeApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: apiCallsInProgress[key].filter((verb) => verb !== value),
    })
  }

  const value = {
    flashProps,
    modalProps,
    apiCallsInProgress,
    setFlashProps,
    setModalProps,
    addApiCall,
    removeApiCall,
  }

  useEffect(() => {
    if (flashProps.hidden === true) return

    flashVisibleSince.current = Number(new Date())

    setTimeout(() => {
      if (Number(new Date()) - flashVisibleSince.current >= 4000) {
        setFlashProps({ ...flashProps, hidden: true })
        flashVisibleSince.current = 0
      }
    }, 4000)
  }, [flashProps])

  return <PageContext.Provider value={value}>{children}</PageContext.Provider>
}

export { PageContext, PageProvider }
