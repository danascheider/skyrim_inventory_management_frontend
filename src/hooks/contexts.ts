import { useContext } from 'react'
import { ColorContext } from '../contexts/colorContext'
import { LoginContext } from '../contexts/loginContext'

const useCustomContext = <T>(cxt: React.Context<T>, msg: string) => {
  const context = useContext(cxt)

  if (!context) throw new Error(msg)

  return context
}

export const useColorScheme = () =>
  useCustomContext(
    ColorContext,
    'useColorScheme must be used within a ColorProvider'
  )

export const useGoogleLogin = () =>
  useCustomContext(
    LoginContext,
    'useGoogleLogin must be used within a LoginProvider'
  )