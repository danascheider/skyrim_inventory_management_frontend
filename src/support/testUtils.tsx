import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { JSDOM } from 'jsdom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const BASE_APP_URI = 'http://localhost:5173'

declare global {
  namespace NodeJS {
    interface Global {
      document: Document
      window: Window
    }
  }
}

const setDom = () => {
  const dom = new JSDOM('<!doctype html><html><body></body></html>', {
    url: 'http://localhost:5173',
  })

  global.window = dom.window as unknown as Window & typeof globalThis
  global.document = dom.window.document
}

/*
 *
 * Global Test Data
 *
 */

export const requireLogin = () => {
  /* noop */
}

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdtoken: () =>
    new Promise<string>((resolve, _reject) => resolve('xxxxxxx')),
} as unknown as User

/*
 *
 * Test Renderers
 *
 */

const renderWithJSDOM = (ui: ReactElement) => {
  setDom()

  return render(ui)
}

export const renderWithRouter = (ui: ReactElement) =>
  renderWithJSDOM(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (ui: ReactElement, authLoading = false) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: testUser,
        token: 'xxxxxxx',
        requireLogin,
        authLoading,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: null,
        token: null,
        authLoading: false,
        requireLogin,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{ user: null, token: null, authLoading: true, requireLogin }}
    >
      {ui}
    </LoginContext.Provider>
  )
