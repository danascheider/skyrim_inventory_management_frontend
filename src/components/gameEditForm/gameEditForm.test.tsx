import { describe, test, expect, vitest } from 'vitest'
import { act, fireEvent } from '@testing-library/react'
import { renderAuthenticated } from '../../support/testUtils'
import { allGames as games } from '../../support/data/games'
import { DONE } from '../../utils/loadingStates'
import { GREEN } from '../../utils/colorSchemes'
import { GamesContext } from '../../contexts/gamesContext'
import { PageProvider } from '../../contexts/pageContext'
import GameEditForm from './gameEditForm'

const contextValue = {
  games,
  gamesLoadingState: DONE,
  createGame: () => {},
  destroyGame: () => {},
}

describe('GameEditForm', () => {
  test('displays the inputs and submit button', () => {
    const game = games[0]

    const wrapper = renderAuthenticated(
      <PageProvider>
        <GamesContext.Provider value={contextValue}>
          <GameEditForm
            gameId={game.id}
            name={game.name}
            description={game.description}
            buttonColor={GREEN}
          />
        </GamesContext.Provider>
      </PageProvider>
    )

    expect(wrapper).toBeTruthy()

    expect(wrapper.getByTestId('nameField')).toBeTruthy()
    expect(wrapper.getByTestId('descriptionField')).toBeTruthy()
    expect(wrapper.getByTestId('submitGameEditForm')).toBeTruthy()
  })

  test('matches snapshot', () => {
    const game = games[0]

    const wrapper = renderAuthenticated(
      <PageProvider>
        <GamesContext.Provider value={contextValue}>
          <GameEditForm
            gameId={game.id}
            name={game.name}
            description={game.description}
            buttonColor={GREEN}
          />
        </GamesContext.Provider>
      </PageProvider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  describe('submitting the form', () => {
    test('calls the update function', () => {
      const game = games[0]

      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={contextValue}>
            <GameEditForm
              gameId={game.id}
              name={game.name}
              description={game.description}
              buttonColor={GREEN}
            />
          </GamesContext.Provider>
        </PageProvider>
      )
    })
  })
})
