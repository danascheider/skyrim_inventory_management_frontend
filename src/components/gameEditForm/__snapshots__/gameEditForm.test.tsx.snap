// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GameEditForm > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_bc52a1"
      >
        <h3
          class="_header_bc52a1"
        >
          Update Game
        </h3>
        <form
          class="_form_bc52a1"
          data-testid="editGame32Form"
          style="--button-background-color: #00a323; --button-text-color: #fff; --button-hover-color: #00921f; --button-border-color: #00821c;"
        >
          <fieldset
            class="_fieldset_bc52a1"
          >
            <label
              class="_label_bc52a1"
              for="name"
            >
              Name
            </label>
            <input
              aria-label="Name"
              class="_input_bc52a1"
              data-testid="nameField"
              name="name"
              pattern="^\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*$"
              title="Name can contain only alphanumeric characters, spaces, commas, hyphens, and apostrophes"
              type="text"
              value="My Game 1"
            />
          </fieldset>
          <fieldset
            class="_fieldset_bc52a1"
          >
            <label
              class="_label_bc52a1"
              for="description"
            >
              Description
            </label>
            <input
              aria-label="Description"
              class="_input_bc52a1"
              data-testid="descriptionField"
              name="description"
              placeholder="Description"
              value="This is a game with a description"
            />
          </fieldset>
          <button
            class="_submit_bc52a1"
            data-testid="submitGameEditForm"
            type="submit"
          >
            Update
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_bc52a1"
    >
      <h3
        class="_header_bc52a1"
      >
        Update Game
      </h3>
      <form
        class="_form_bc52a1"
        data-testid="editGame32Form"
        style="--button-background-color: #00a323; --button-text-color: #fff; --button-hover-color: #00921f; --button-border-color: #00821c;"
      >
        <fieldset
          class="_fieldset_bc52a1"
        >
          <label
            class="_label_bc52a1"
            for="name"
          >
            Name
          </label>
          <input
            aria-label="Name"
            class="_input_bc52a1"
            data-testid="nameField"
            name="name"
            pattern="^\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*$"
            title="Name can contain only alphanumeric characters, spaces, commas, hyphens, and apostrophes"
            type="text"
            value="My Game 1"
          />
        </fieldset>
        <fieldset
          class="_fieldset_bc52a1"
        >
          <label
            class="_label_bc52a1"
            for="description"
          >
            Description
          </label>
          <input
            aria-label="Description"
            class="_input_bc52a1"
            data-testid="descriptionField"
            name="description"
            placeholder="Description"
            value="This is a game with a description"
          />
        </fieldset>
        <button
          class="_submit_bc52a1"
          data-testid="submitGameEditForm"
          type="submit"
        >
          Update
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
